readparameters.h

```diff
@@ -22,57 +22,169 @@
 
 #ifndef READPARAMETERS_H
 #define READPARAMETERS_H
+
+#include <boost/lexical_cast.hpp>
+#include <boost/program_options.hpp>
+#include <fstream>
+#include <iomanip>
+#include <iostream>
 #include <limits>
 #include <mpi.h>
 #include <stdint.h>
 #include <string>
+#include <typeinfo>
 #include <vector>
 
 #include "common.h"
+#include "version.h"
+
+namespace PO = boost::program_options;
+
+class Readparameters {
+public:
+   Readparameters(int cmdargc, char* cmdargv[]);
+   ~Readparameters();
+
+   /** Add a new input parameter.
+    * Note that parse must be called in order for the input file(s) to be re-read.
+    * Only called by the root process.
+    * @param name The name of the parameter, as given in the input file(s).
+    * @param desc Description for the parameter.
+    * @param defValue Default value for variable.
+    */
+   static void add(const std::string& name, const std::string& desc, const std::string& defValue) {
+      int rank;
+      MPI_Comm_rank(MPI_COMM_WORLD, &rank);
+      if (rank == MASTER_RANK) {
+         options[name] = "";
+         isOptionParsed[name] = false;
+         descriptions->add_options()(name.c_str(), PO::value<std::string>(&(options[name]))->default_value(defValue),
+                                     desc.c_str());
+      }
+   }
+
+   template <typename T> static void add(const std::string& name, const std::string& desc, const T& defValue) {
```

Comment:
I'm wondering (without having tested): Does this need a separate specialization for ```const char *```? In the current implementation, there's the issue that ```const char *``` is treated as a numeric type, since it can be typeconverted to one. Hence, their string parameter values are typeconverted to a number, which ends up being 0, so the string becomes a nullpointer at runtime.

The "proper" way to use string parameters is ```RP::get("name", "description", std::string("default"));``` at the moment. I wondr if a simple extra specialization would fix this.

Reply:
I haven't seen an example for parsing character type. Perhaps it would be better to treat everything as string first and then do the type conversion. If we want to test this then perhaps we should add a test for this edge case. For now I will ignore this potential issue.